plugins/org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/JavaDerivedStateComputer.xtend

```diff
@@ -0,0 +1,157 @@
+		val result = parser.dietParse(compilationUnit, compilationResult)
+		for (type : result.types) {
+			val packageName = result.currentPackage.importName.map[String.valueOf(it)].join('.')
+			val jvmType = createType(type, packageName)
+			resource.contents.add(jvmType)
+		}
+	}
+	
+	def JvmDeclaredType createType(TypeDeclaration type, String packageName) {
+		val jvmType = switch (TypeDeclaration.kind(type.modifiers)) {
+			case TypeDeclaration.CLASS_DECL :
+				TypesFactory.eINSTANCE.createJvmGenericType
+			case TypeDeclaration.INTERFACE_DECL :
+				TypesFactory.eINSTANCE.createJvmGenericType => [
+					interface = true
+				]
+			case TypeDeclaration.ENUM_DECL :
+				TypesFactory.eINSTANCE.createJvmEnumerationType
+			case TypeDeclaration.ANNOTATION_TYPE_DECL :
+				TypesFactory.eINSTANCE.createJvmAnnotationType
+		}
+		jvmType.packageName = packageName
+		jvmType.simpleName = String.valueOf(type.name)
+		if (jvmType instanceof JvmGenericType) {
+			if (type.typeParameters != null) {
+				for (typeParam : type.typeParameters) {
+					val jvmTypeParam = TypesFactory.eINSTANCE.createJvmTypeParameter
+					jvmTypeParam.name = String.valueOf(typeParam.name)
+					jvmType.typeParameters += jvmTypeParam
+				}
+			}
+		}
+		if (type.memberTypes != null) {
+			for (nestedType : type.memberTypes) {
+				val nested = createType(nestedType, null)
+				jvmType.members += nested
+			}
+		}
+		return jvmType
+	}
+	
+	//TODO make this return type inferred and do a full build in this project to see a bug
+	def ICompilationUnit getCompilationUnit(Resource resource) {
+		(resource as JavaResource).getCompilationUnit()
+	}
+	
+	def void installFull(Resource resource) {
+		val compilationUnit = getCompilationUnit(resource)
+		val classLoader = getClassLoader(resource)
+		val data = ResourceDescriptionsData.ResourceSetAdapter.findResourceDescriptionsData(resource.resourceSet)
+		val nameEnv = new IndexAwareNameEnvironment(classLoader, data, stubGenerator)
+		val compiler = new Compiler(nameEnv, DefaultErrorHandlingPolicies.proceedWithAllProblems(), getCompilerOptions(), [
+			if (Arrays.equals(it.fileName, compilationUnit.fileName)) {
+				val map = newHashMap
+				var List<String> topLevelTypes = newArrayList
+				for (cf : it.getClassFiles()) {
+					val className = cf.compoundName.map[String.valueOf(it)].join('.')
+					map.put(className, cf.bytes)
+					if (!cf.isNestedType) {
+						topLevelTypes += className
+					}
+				}
+				val inMemClassLoader = new InMemoryClassLoader(map, classLoader)
+				for (topLevel : topLevelTypes) {
+					val builder = new JvmDeclaredTypeBuilder(new BinaryClass(topLevel, inMemClassLoader), new ClassFileBytesAccess(), inMemClassLoader)
+					resource.contents += builder.buildType
+				}
+			}
+		], new DefaultProblemFactory())
+		compiler.compile(#[compilationUnit])
+	}
+	
+	protected def CompilerOptions getCompilerOptions() {
+		val jdkVersion = ClassFileConstants.JDK1_7;
```

Comment:
The compiler options will need to be configured from outside. They have to match whatever the project uses. This is just a default for the time being.